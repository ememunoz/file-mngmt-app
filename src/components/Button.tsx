import React, { useState } from 'react'
import { Pressable, StyleSheet, Text } from 'react-native'
import type { GestureResponderEvent, StyleProp, ViewStyle } from 'react-native'

type Props = {
  onPress: (event: GestureResponderEvent) => void
  children: string
  style?: ViewStyle | ViewStyle[]
}

const Button = ({ children, onPress, style: customStyle }: Props) => {
  const [isPress, setIsPress] = useState(false)

  const handlePressIn = () => setIsPress(true)
  const handlePressOut = () => setIsPress(false)

  let finalStyle: StyleProp<ViewStyle>[]
  if (Array.isArray(customStyle)) {
    finalStyle = [styles.button, ...customStyle]
  } else {
    finalStyle = [styles.button, customStyle]
  }
  if (isPress) {
    finalStyle.push(styles.buttonPress)
  }

  return (
    <Pressable
      style={finalStyle}
      onPress={onPress}
      onPressIn={handlePressIn}
      onPressOut={handlePressOut}
    >
      <Text style={styles.text}>{children}</Text>
    </Pressable>
  )
}

const styles = StyleSheet.create({
  button: {
    alignItems: 'center',
    justifyContent: 'center',
    paddingVertical: 12,
    paddingHorizontal: 32,
    borderRadius: 4,
    elevation: 3,
    backgroundColor: 'hsl(243, 100%, 69%)',
  },
  buttonPress: {
    elevation: 0,
    backgroundColor: 'hsl(243, 80%, 71%)',
  },
  text: {
    fontSize: 16,
    lineHeight: 21,
    fontWeight: 'bold',
    letterSpacing: 0.25,
    color: 'white',
  },
})

export default Button
