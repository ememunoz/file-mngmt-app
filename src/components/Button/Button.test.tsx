import { fireEvent, render, screen } from '@testing-library/react-native'
import React from 'react'

import Button from './Button'

describe('<Button />', () => {
  it('renders correctly', () => {
    render(<Button onPress={() => {}}>Tap me</Button>)
    expect(screen.toJSON()).toMatchSnapshot()
  })

  it('calls onPress when pressed', () => {
    const mockFunction = jest.fn()
    render(<Button onPress={mockFunction}>Tap me</Button>)
    expect(mockFunction.mock.calls.length).toBe(0)
    fireEvent.press(screen.getByText('Tap me'))
    expect(mockFunction.mock.calls.length).toBe(1)
  })
})
