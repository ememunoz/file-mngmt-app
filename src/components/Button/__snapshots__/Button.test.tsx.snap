// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "hsl(243, 100%, 69%)",
        "borderRadius": 4,
        "elevation": 3,
        "justifyContent": "center",
        "paddingHorizontal": 32,
        "paddingVertical": 12,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      Object {
        "color": "white",
        "fontSize": 16,
        "fontWeight": "bold",
        "letterSpacing": 0.25,
        "lineHeight": 21,
      }
    }
  >
    Tap me
  </Text>
</View>
`;
